testkit/src/main/scala/ch/epfl/scala/bsp/testkit/client/TestClient.scala

```diff
@@ -0,0 +1,433 @@
+package ch.epfl.scala.bsp.testkit.client
+
+import java.io.File
+import java.util.Optional
+import java.util.concurrent.CompletableFuture
+
+import ch.epfl.scala.bsp.testkit.client.mock.{MockCommunications, MockSession}
+import ch.epfl.scala.bsp4j._
+
+import scala.annotation.tailrec
+import scala.collection.convert.ImplicitConversions.`collection asJava`
+import scala.collection.mutable
+import scala.compat.java8.FutureConverters._
+import scala.compat.java8.OptionConverters._
+import scala.concurrent.Await
+import scala.concurrent.duration.{Duration, DurationInt}
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Try}
+
+case class TestClient(
```

Comment:
case class not a good fit here, it usually implies that it's an immutable data holder, whereas this class kind of holds state

Reply:
Thanks for noticing this. The initial plan for the class was different but I didn't remember to change this